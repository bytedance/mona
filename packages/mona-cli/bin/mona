#! /usr/bin/env node

const mona = require('../dist').default;
mona();
