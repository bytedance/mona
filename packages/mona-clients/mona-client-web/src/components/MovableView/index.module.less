.touch {
  display: inline-block;
  will-change: transform;
  transform-origin: left top;
}
