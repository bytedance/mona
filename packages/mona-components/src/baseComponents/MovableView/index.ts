import React from 'react';
import createComponent from '../../createComponent';
import { BaseProps, EventHandler } from '../base';

export interface MovableViewProps extends BaseProps {
  direction?: 'all' | 'vertical' | 'horizontal' | 'none';
  inertia?: boolean;
  outOfBounds?: boolean;
  x?: string | number;
  y?: string | number;
  damping?: number;
  friction?: number;
  disabled?: boolean;
  scale?: boolean;
  scaleMin?: number;
  scaleMax?: number;
  scaleValue?: number;
  animation?: boolean;
  onChange?: EventHandler;
  onScale?: EventHandler;
  onHtouchMove?: EventHandler;
  onVtouchMove?: EventHandler;
}

export const MovableView: React.ComponentType<MovableViewProps> = createComponent<MovableViewProps>('movable-view')
