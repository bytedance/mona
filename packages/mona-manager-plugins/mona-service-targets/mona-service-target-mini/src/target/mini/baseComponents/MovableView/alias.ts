import { bubbleEventsAlias, basePropAlias, propAliasMap } from '../prop';
import { MovableViewAlias as Alias } from '../type';

const alias: Alias = {
  ...basePropAlias,
  ...bubbleEventsAlias,
  direction: propAliasMap.direction,
  inertia: propAliasMap.inertia,
  outOfBounds: propAliasMap.outOfBounds,
  x: propAliasMap.x,
  y: propAliasMap.y,
  damping: propAliasMap.damping,
  friction: propAliasMap.friction,
  disabled: propAliasMap.disabled,
  scale: propAliasMap.scale,
  scaleMin: propAliasMap.scaleMin,
  scaleMax: propAliasMap.scaleMax,
  scaleValue: propAliasMap.scaleValue,
  animation: propAliasMap.animation,
  onChange: propAliasMap.onChange,
  onHtouchMove: propAliasMap.onHtouchMove,
  onScale: propAliasMap.onScale,
  onVtouchMove: propAliasMap.onVtouchMove,
};

export default alias;
