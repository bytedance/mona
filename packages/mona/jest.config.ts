import baseConfig from '../../jestconfig.base';

module.exports = baseConfig;
