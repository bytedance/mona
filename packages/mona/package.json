{
  "name": "@bytedance/mona",
  "version": "0.2.0-alpha.9",
  "description": "Merchant app's developing and building tools",
  "main": "./dist/index",
  "scripts": {
    "build": "yarn clean && ttsc -b",
    "start": "ttsc -b -w",
    "clean": "rm -rf dist",
    "test": "jest"
  },
  "files": [
    "dist"
  ],
  "author": "zhangxiaowei.chris@bytedance.com",
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "devDependencies": {
    "copy-webpack-plugin": "^10.2.0",
    "react": "^17.0.2",
    "webpack": "^5.58.2",
    "webpack-chain": "^6.5.1"
  },
  "peerDependencies": {
    "react": "^17.0.2"
  },
  "gitHead": "cc5aa5283af595fd7d8a7b4921f54e2797504edd"
}
