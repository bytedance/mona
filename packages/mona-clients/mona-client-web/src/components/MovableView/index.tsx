import React, { useRef, useEffect, useState, useCallback, TouchEventHandler } from 'react';
import { MovableViewProps } from '@bytedance/mona';
import { formatTouchEvent, genEvent } from '../utils';
import cs from 'classnames';
import styles from './index.module.less';

type Position = { x: number; y: number };
const getDistance = (start: Position, stop: Position) => {
  return Math.hypot(stop.x - start.x, stop.y - start.y);
};

const checkPosi = (
  value: Position,
  wrapperInfo: { wrapperWidth: number; wrapperHeight: number },
  itemInfo: { itemWidth: number; itemHeight: number },
  outOfBounds: boolean,
) => {
  let { x, y } = value;
  const { wrapperWidth, wrapperHeight } = wrapperInfo;
  const { itemWidth, itemHeight } = itemInfo;
  let [overflowX, overflowY] = [true, true];
  const xLimit = wrapperWidth - itemWidth;
  const yLimit = wrapperHeight - itemHeight;

  let [xMin, xMax] = [0, xLimit];
  let [yMin, yMax] = [0, yLimit];

  if (outOfBounds) {
    const xOverflowLimit = (itemWidth / 2) * 0.8;
    const yOverflowLimit = (itemHeight / 2) * 0.8;

    xMin -= xOverflowLimit;
    xMax += xOverflowLimit;

    yMin -= yOverflowLimit;
    yMax += yOverflowLimit;
  }

  if (x < xMin) {
    x = xMin;
  } else if (x > xMax) {
    x = xMax;
  }
  if (x >= 0 && x <= xLimit) {
    overflowX = false;
  }
  if (y < yMin) {
    y = yMin;
  } else if (y > yMax) {
    y = yMax;
  }
  if (y >= 0 && y <= yLimit) {
    overflowY = false;
  }
  return {
    data: { x, y },
    overflow: overflowY || overflowX,
  };
};
const MovableView: React.FC<MovableViewProps> = props => {
  let {
    // scale,
    scaleValue = 1,
    scaleMax = 10,
    scaleMin = 0.5,
    children,
    disabled,
    x = 0,
    y = 0,
    // onScale,
    damping = 20,
    friction = 2,
    //@ts-ignore
    wrapperWidth,
    //@ts-ignore
    wrapperHeight,
    //@ts-ignore
    wrapperRef,
    //@ts-ignore

    scaleArea,
    className,
  } = props;
  // const ref = useRef<HTMLDivElement>(null);
  const domRef = useRef<HTMLSpanElement>(null);

  const disRef = useRef({ distX: 0, distY: 0 });
  const propsRef = useRef<Required<MovableViewProps> & Record<string, any>>(props as Required<MovableViewProps>);
  const stateRef = useRef<Record<string, any>>({});

  const [{ x: innerX, y: innerY }, setPosition] = useState<Position>({ x: 0, y: 0 });
  const [executing, setExecuting] = useState(false);

  const [innerScale, setScale] = useState(scaleValue);

  useEffect(() => {
    const sValue = scaleValue < scaleMax ? (scaleValue > scaleMin ? scaleValue : scaleMin) : scaleMax;
    setScale(sValue);
  }, [scaleValue, scaleMax, scaleMin]);

  propsRef.current = props as Required<MovableViewProps>;
  stateRef.current = {
    executing,
  };

  const setPositionWithCheck = useCallback(
    (posi: Position | ((value: Position) => Position), outOfBounds?: boolean, e?: React.TouchEvent) => {
      outOfBounds = outOfBounds === undefined ? propsRef.current.outOfBounds : outOfBounds;
      const check = (posi: Position) => {
        const { width: realWidth, height: realHeight } = domRef.current!.getBoundingClientRect() ?? {};
        const { data, overflow } = checkPosi(
          posi,
          { wrapperWidth, wrapperHeight },
          {
            // 获取scale之后的宽高
            itemHeight: realHeight,
            itemWidth: realWidth,
          },
          !!outOfBounds,
        );
        let formatEvent;
        if (e) {
          formatEvent = formatTouchEvent({ event: e, type: 'change' });
          formatEvent.detail = { ...data };
          if (overflow) {
            formatEvent.detail.source = 'touch-out-of-bounds';
          } else {
            formatEvent.detail.source = 'touch';
          }
        } else {
          formatEvent = genEvent({ detail: {}, type: 'change' });
          formatEvent.detail = { ...data };
          formatEvent.detail.source = '';
        }

        propsRef.current.onChange?.(formatEvent);

        return { data, overflow };
      };
      if (typeof posi === 'function') {
        setPosition(pre => check(posi(pre)).data);
      } else {
        setPosition(check(posi).data);
      }
    },
    [wrapperWidth, wrapperHeight],
  );

  useEffect(() => {
    setExecuting(true);
    setPositionWithCheck({ x: +x, y: +y }, false);
  }, [x, y, setPositionWithCheck]);

  const handleTouchStart = useCallback<TouchEventHandler<HTMLSpanElement>>(e => {
    setExecuting(false);
    const { clientX, clientY } = e.targetTouches[0];
    const posInfo = domRef.current!.getBoundingClientRect();
    disRef.current.distX = clientX - posInfo.left;
    disRef.current.distY = clientY - posInfo.top;
  }, []);
  const handleScale = useCallback<TouchEventHandler<HTMLSpanElement>>(e => {
    const { clientX, clientY } = e.targetTouches[0];
    const { distX, distY } = disRef.current;
    const L = clientX - distX,
      T = clientY - distY;
    if (!propsRef.current.scale) {
      return;
    }
    const touches = e.targetTouches;
    const events = touches[0];
    const events2 = touches[1];
    const zoom = getDistance(
      {
        x: events.pageX,
        y: events.pageY,
      },
      {
        x: events2.pageX,
        y: events2.pageY,
      },
    );

    setScale(preScale => {
      const { scaleMin, scaleMax } = propsRef.current;
      const newScaleValue = preScale * zoom;
      const res = newScaleValue < scaleMax ? (newScaleValue > scaleMin ? newScaleValue : scaleMin) : scaleMax;
      const formatEvent = formatTouchEvent({ event: e, type: 'change' });
      formatEvent.detail = {
        x: L,
        y: T,
        scale: res,
      };
      propsRef.current.onScale?.(formatEvent);

      return res;
    });
  }, []);
  const handleTouchMove = useCallback<TouchEventHandler<HTMLSpanElement>>(
    e => {
      const tLength = e.targetTouches.length;
      const { clientX, clientY } = e.targetTouches[0];
      const { distX, distY } = disRef.current;
      const L = clientX - distX,
        T = clientY - distY;
      if (tLength === 1) {
        const { direction } = propsRef.current;
        if (direction === 'horizontal') {
          setPositionWithCheck(pre => ({ x: L, y: pre.y }), undefined, e);
        } else if (direction === 'vertical') {
          setPositionWithCheck(pre => ({ x: pre.x, y: T }), undefined, e);
        } else if (direction === 'all') {
          setPositionWithCheck({ x: L, y: T }, undefined, e);
        } else {
          return;
        }
      } else if (tLength === 2) {
        !scaleArea && handleScale(e);
      }
    },
    [setPositionWithCheck, handleScale, scaleArea],
  );
  const handleTouchEnd = useCallback<TouchEventHandler<HTMLSpanElement>>(
    e => {
      setPosition(({ x, y }) => {
        const { wrapperWidth, wrapperHeight } = propsRef.current;
        const { width: realWidth, height: realHeight } = domRef.current!.getBoundingClientRect() ?? {};

        const { data } = checkPosi(
          { x, y },
          { wrapperWidth, wrapperHeight },
          {
            itemHeight: realHeight,
            itemWidth: realWidth,
          },
          false,
        );
        let formatEvent;
        formatEvent = formatTouchEvent({ event: e, type: 'change' });
        formatEvent.detail = { ...data };
        formatEvent.detail.source = 'out-of-bounds';
        if (data.x !== x || data.y !== y) {
          propsRef.current.onChange?.(formatEvent);
          setExecuting(true);
        }

        return data;
      });
    },
    [setPositionWithCheck],
  );
  const animationTime = `${4 / friction / damping}s`;

  const transitionStyle = executing
    ? {
        transition: `scale ${animationTime}, transform ${animationTime}`,
      }
    : { transition: `scale ${animationTime}` };

  useEffect(() => {
    const wrapperDom = wrapperRef.current as HTMLDivElement;
    if (wrapperDom && scaleArea) {
      wrapperDom.addEventListener('touchmove', handleScale as any);
    }
    return () => {
      scaleArea && wrapperDom.removeEventListener('touchmove', handleScale as any);
    };
  }, [handleScale, scaleArea]);

  return (
    //@ts-ignore
    <span
      ref={domRef}
      //@ts-ignore
      onTransitionEnd={() => setExecuting(false)}
      //@ts-ignore
      onTouchStart={disabled ? undefined : handleTouchStart}
      //@ts-ignore
      onTouchMove={disabled ? undefined : handleTouchMove}
      //@ts-ignore
      onTouchEnd={disabled ? undefined : handleTouchEnd}
      style={{
        transform: `translate(${innerX}px, ${innerY}px) scale(${innerScale})`,
        ...transitionStyle,
      }}
      className={cs(className, styles.touch)}
    >
      {children}
    </span>
  );
};

export default MovableView;

MovableView.defaultProps = {
  direction: 'none',
  inertia: false,
  outOfBounds: false,
  damping: 20,
  friction: 2,
  disabled: false,
  scale: false,
  scaleMin: 0.5,
  scaleMax: 10,
  animation: true,
};
