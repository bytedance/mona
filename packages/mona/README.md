# @bytedance/mona
对外暴露提供构建时API
## 配置创建
### createProjectConfig
创建项目配置，`mona.config.ts`中使用
#### 例子
```js
// mona.config.ts
import { createProjectConfig } from '@bytedance/mona';
export default createProjectConfig({
  projectName: 'demo',
  input: './src/app.tsx',
  output: 'dist'
})
```

#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| projectName  | 应用/插件名称 | 是 | string | - |
| input  | 入口文件路径 | 否 | string | - |
| output  | 打包后的文件目录路径 | 否 | string | - |
| raw  | 定义函数来修改webpack配置 | 否 | (options: any) => any | - |
| dev  | 用来修改本地服务配置 | 否 | string | { port: 9999 } |

### createAppConfig
创建应用配置，`app.config.ts`中使用
#### 例子
```js
// app.config.ts
import { createAppConfig } from '@bytedance/mona';
export default createAppConfig({
  pages: [
    'pages/Home/index',
    'pages/Info/index',
    'pages/List/index'
  ]
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| pages  | 页面路径 | 是 | string[] | - |

默认`pages`的第一项为首页
### createPageConfig
创建页面配置，需在每个页面目录的`page.config.ts`中使用

#### 例子
```js
// page.config.ts
import { createPageConfig } from '@bytedance/mona';
export default createPageConfig({
  navigationBarTitleText: 'page title'
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| navigationBarTitleText  | 当前页面标题 | 否 | string | - |
