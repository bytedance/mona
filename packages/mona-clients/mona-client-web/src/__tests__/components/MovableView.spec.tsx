import React from 'react';
import { mount } from 'enzyme';
import { configure } from 'enzyme';
import MovableView from '../../components/MovableView';
import mountTest from '../../../../../tests/shared/mountTest';
// import handlerTest from '../../../../../tests/shared/handlerTest';
import { MovableViewProps } from '@bytedance/mona';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

configure({ adapter: new Adapter() });


describe('web component: MovableView', () => {
  mountTest(MovableView)
  // handlerTest({ Component: MovableView, firstTag: 'span',  });

  it('should render correctly', () => {
    expect(mount(<MovableView />).render()).toMatchSnapshot();
  })
  
  it('should have right properties', () => {
    const props: MovableViewProps = {
      direction: 'all',
      inertia: true,
      outOfBounds: true,
      x: 10,
      y: 10,
      damping: 10,
      friction: 10,
      disabled: true,
      scale: true,
      scaleMin: 1,
      scaleMax: 2,
      scaleValue: 4,
      animation: true,
      className: 'class-name',
      style: { color: 'red' },
    }
    const wrapper = mount(<MovableView {...props}>hello</MovableView>);
    expect(wrapper.render()).toMatchSnapshot();
  })
})
