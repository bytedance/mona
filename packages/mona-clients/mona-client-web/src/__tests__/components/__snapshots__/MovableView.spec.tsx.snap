// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web component: MovableView should have right properties 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "class-name touch",
      "direction": "all",
      "style": "transform: translate(10px, 10px) scale(2); transition: scale 0.04s, transform 0.04s;",
    },
    "children": Array [
      Node {
        "data": "hello",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "span",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "direction": undefined,
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "direction": undefined,
      "style": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`web component: MovableView should render correctly 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "touch",
      "direction": "none",
      "style": "transform: translate(0px, 0px) scale(1); transition: scale 0.1s, transform 0.1s;",
    },
    "children": Array [],
    "name": "span",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "direction": undefined,
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "direction": undefined,
      "style": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
